plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs

```diff
@@ -315,7 +316,7 @@ maybeAddSubst a s n rhs = do
 
     -- Check whether we've been told specifically to inline this
     hints <- asks _hints
-    let hinted = shouldInline hints a n
+    let hinted = shouldInline hints a n (termAnn rhs')
```

Comment:
Also, this is now inconsistent with the UPLC inliner!

Reply:
So there are two reasons why we look at both the binding annotation and the RHS annotation:
- Some inlining criteria (such as whether the RHS is `trace`) depends only on the RHS. So it's more convenient to just annotate the RHS with that, rather than having to propagate that annotation to the entire binding.
- The binding may already have other annotations like `DatatypeComponent`. To have additional annotations on the binding require a bigger change than this.

Does this make sense?

btw I made the same change to the UPLC inliner.

Reply:
> Some inlining criteria (such as whether the RHS is trace) depends only on the RHS

I disagree with this framing, I think. I think it really is a property of the binding: we want to inline _that binding_ not _that term_. For example:

```
let
   -- Original binding that we want to inline 
   trace = ...
   -- Just some other binding
   trace2 = trace
in trace2 ...
```

Should we inline `trace2`? I think we should not, but with your implementation we will, because the "should inline" property attaches to the term, not the binding.

> The binding may already have other annotations like DatatypeComponent. To have additional annotations on the binding would probably require a bigger change than this.

The `Provenance` annotations all have a "parent" provenance, which should eventually top out with an `Original` provenance. That is, we should be able to write a `getOriginal :: Provenance a -> a` function (okay, maybe we need to do something for `MultipleSources`, either return `Nothing` or require `Monoid a`, I guess). So we should be able to pull out the underlying annotation in all cases!

Reply:
@michaelpj 

> Should we inline `trace2`?

That's exactly what I'd expect, so

> I think we should not

why not?

When I just started reviewing this, I thought the same as you, however after some staring at the code I realized I like what's implemented here much more. Not only does it allow us to inline the obvious candidates like `/\a -> \text x -> trace {a} text x`, but also their instantiations like `\text (x : integer) -> trace {integer} text x` and the partial application `\(x : integer) -> trace {integer} text x` too. This means one can inline a term, then partially reduce it as its use site, inline it, compute a little more, opening another opportunity for inlining, and so forth.

It's much trickier though as we now can end up inlining closures. What if `text` is some huge chunk of text, do we then want to inline `\(x : integer) -> trace {integer} text x`? Probably not. BTW, is it ensured that we only inline `trace` when traces are ignored or is it considered fine to always inline them?

But currently we have <10 `AnnInline` mentions, so I wouldn't worry too much about the complexity of the approach. It's certainly not any more complex than any other pass that we have.

Reply:
The goal here is somewhat limited: we're going to make a specific binding which doesn't look like it should inline, and we want it to inline (including the type/lambda abstractions inside it). 

> But currently we have <10 AnnInline mentions, so I wouldn't worry too much about the complexity of the approach. It's certainly not any more complex than any other pass that we have.

Well, I think it's making it more complex than it was before. Previously we just looked at bindings, now we also look at terms!

Reply:
@michaelpj Ok, so to propagate annotations from the RHS of a binding to the binding itself, we need to do these:
- Update `runDefT`: #4637  (I think we should do this regardless, hence the PR)
- Every time we construct a `TermBind` we must remember to propagate the annotation. This can be done by a smart constructor so is not too bad, but is there really no case where we _don't_ want to propagate?
- Whenever we simplify RHS, turning `let v = RHS in expr` into `let v = RHS' in expr`, if `RHS'` has a new annotation, we must remember to modify `v`'s annotation.

The last one sounds pretty unattractive. What do you think?

Reply:
Okay, here's what I was thinking:
- In `Definitions`, propagate the annotation on the `VarDecl` to the `Binding` also. We're quite fuzzy about this, but I think hanging it off the variable is reasonable.
- Then we just need to make sure that the `VarDecl`s we create have the bindings that we want. That shouldn't be too bad: we can just pass an annotation into `defineBuiltinTerm` etc. and use that.

So basically I think we should keep the strategy of "indicate what annotation to use by attaching it to the data in `Definitions`" rather than "pass in functions at the end to fix up the annotations", but I think there is a good place to put the annotation.